tpm/tss2/signer.go

```diff
@@ -0,0 +1,209 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tss2
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/google/go-tpm/legacy/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var primaryParams = tpm2.Public{
+	Type:       tpm2.AlgECC,
+	NameAlg:    tpm2.AlgSHA256,
+	Attributes: tpm2.FlagStorageDefault,
+	ECCParameters: &tpm2.ECCParams{
+		Symmetric: &tpm2.SymScheme{
+			Alg:     tpm2.AlgAES,
+			KeyBits: 128,
+			Mode:    tpm2.AlgCFB,
+		},
+		Sign: &tpm2.SigScheme{
+			Alg: tpm2.AlgNull,
+		},
+		CurveID: tpm2.CurveNISTP256,
+	},
+}
+
+// Signer implements [crypto.Signer] using a [TPMKey].
+type Signer struct {
+	rw        io.ReadWriter
+	publicKey crypto.PublicKey
+	tpmKey    *TPMKey
+}
+
+// CreateSigner creates a new [crypto.Signer] with the given TPM (rw) and
+// [TPMKey]. The caller is responsible of opening and closing the TPM.
+func CreateSigner(rw io.ReadWriter, key *TPMKey) (crypto.Signer, error) {
+	switch {
+	case rw == nil:
+		return nil, fmt.Errorf("invalid TPM channel: rw cannot be nil")
+	case !key.Type.Equal(oidLoadableKey):
+		return nil, fmt.Errorf("invalid TSS2 key: type %q is not valid", key.Type.String())
+	case len(key.Policy) != 0:
+		return nil, errors.New("invalid TSS2 key: policy is not implemented")
+	case len(key.AuthPolicy) != 0:
+		return nil, errors.New("invalid TSS2 key: auth policy is not implemented")
+	case len(key.Secret) > 0:
+		return nil, errors.New("invalid TSS2 key: secret should not be set")
+	case !validateParent(key.Parent):
+		return nil, fmt.Errorf("invalid TSS2 key: parent '%d' is not valid", key.Parent)
+	case !validateKey(key.PublicKey):
+		return nil, errors.New("invalid TSS2 key: public key is invalid")
+	case !validateKey(key.PrivateKey):
+		return nil, errors.New("invalid TSS2 key: private key key is invalid")
+	}
+
+	public, err := tpm2.DecodePublic(key.PublicKey[2:])
+	if err != nil {
+		return nil, fmt.Errorf("error decoding public key: %w", err)
+	}
+	if public.Type != tpm2.AlgRSA && public.Type != tpm2.AlgECC {
+		return nil, fmt.Errorf("invalid TSS2 key: public key type %q is not valid", public.Type.String())
+	}
+	publicKey, err := public.Key()
+	if err != nil {
+		return nil, fmt.Errorf("error creating public key: %w", err)
+	}
+
+	return &Signer{
+		rw:        rw,
+		publicKey: publicKey,
+		tpmKey:    key,
+	}, nil
+}
+
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+func (s *Signer) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	parentHandle := tpmutil.Handle(s.tpmKey.Parent)
+	if !handleIsPersistent(s.tpmKey.Parent) {
+		parentHandle, _, err = tpm2.CreatePrimary(s.rw, parentHandle, tpm2.PCRSelection{}, "", "", primaryParams)
```

Comment:
Might be a nice opportunity to not use the legacy TPM2 APIs in this new package. Not a blocker to get this merged, though.

Reply:
Probably a different PR.